desc/integrals/basis.py

```diff
@@ -58,6 +85,621 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
+        #      ∂f/∂y =      ∑ₙ₌₀ᴺ⁻¹ aₙ(x) n Uₙ₋₁(y)
+        # sign ∂f/∂y = sign ∑ₙ₌₀ᴺ⁻¹ aₙ(x) n sin(n arcos y)
+        df_dy_sign = jnp.sign(
+            jnp.linalg.vecdot(
+                n * jnp.sin(n * jnp.arccos(y)[..., jnp.newaxis]),
+                self.cheb[..., jnp.newaxis, :],
+            )
+        )
+        y = bijection_from_disc(y, self.domain[0], self.domain[-1])
+        return y, is_intersect, df_dy_sign
+
+    def intersect1d(self, k=0.0, *, num_intersect=None, pad_value=0.0):
+        """Coordinates z(x, yᵢ) such that fₓ(yᵢ) = k for every x.
+
+        Parameters
+        ----------
+        k : jnp.ndarray
+            Shape must broadcast with (..., *cheb.shape[:-2]).
+            Specify to find solutions yᵢ to fₓ(yᵢ) = k. Default 0.
+        num_intersect : int or None
+            Specify to return the first ``num_intersect`` intersects.
+            This is useful if ``num_intersect`` tightly bounds the actual number.
+
+            If not specified, then all intersects are returned. If there were fewer
+            intersects detected than the size of the last axis of the returned arrays,
+            then that axis is padded with ``pad_value``.
+        pad_value : float
+            Value with which to pad array. Default 0.
+
+        Returns
+        -------
+        z1, z2 : tuple[jnp.ndarray]
+            Shape broadcasts with (..., *self.cheb.shape[:-2], num intersect).
+            Tuple of length two (z1, z2) of coordinates of intersects.
+            The points are ordered and grouped such that the straight line path
+            between ``z1`` and ``z2`` resides in the epigraph of f.
+
+        """
+        errorif(
+            self.Y < 2,
+            NotImplementedError,
+            "This method requires a Chebyshev spectral resolution of Y > 1, "
+            f"but got Y = {self.Y}.",
+        )
+
+        # Add axis to use same k over all Chebyshev series of the piecewise spline.
+        y, is_intersect, df_dy_sign = self.intersect2d(
+            jnp.atleast_1d(k)[..., jnp.newaxis]
+        )
+        # Flatten so that last axis enumerates intersects along the piecewise spline.
+        y, is_intersect, df_dy_sign = map(
+            flatten_matrix, (self._isomorphism_to_C1(y), is_intersect, df_dy_sign)
+        )
+
+        # Note for bounce point applications:
+        # We ignore the degenerate edge case where the boundary shared by adjacent
+        # polynomials is a left intersection i.e. ``is_z1`` because the subset of
+        # pitch values that generate this edge case has zero measure. By ignoring
+        # this, for those subset of pitch values the integrations will be done in
+        # the hypograph of |B|, which will yield zero. If in future decide to
+        # not ignore this, note the solution is to
+        # 1. disqualify intersects within ``_eps`` from ``domain[-1]``
+        # 2. Evaluate sign in ``intersect2d`` at boundary of Chebyshev polynomial
```

Comment:
the reason for ignoring the edge case is detecting them itself is a non-differentiable operation?

Reply:
Yes arcos(+-1) has no derivative. You can detect them differentiable if you loop through every element in the array and add a jax.conditional to evaluate with a different formula if the value is +-1. This is clearly not worth it.